sockeye/encoder.py

```diff
@@ -413,3 +470,173 @@ def get_rnn_cells(self) -> List[mx.rnn.BaseRNNCell]:
+        """
+
+        total_num_filters = sum(self.num_filters)
+        encoded_seq_len = self.get_encoded_seq_len(seq_len)
+
+        # (batch_size, channel=1, seq_len, num_embed)
+        data = mx.sym.Reshape(data=data, shape=(-1, 1, seq_len, self.num_embed))
+
+        # Convolution filters of width 1..N
+        conv_outputs = []
+        for filter_width, num_filter in enumerate(self.num_filters, 1):
+            # "half" padding: output length == input length
+            pad_before = ceil((filter_width - 1) / 2)
+            pad_after = floor((filter_width - 1) / 2)
+            # (batch_size, channel=1, seq_len + (filter_width - 1), num_embed)
+            padded = mx.sym.pad(data=data,
+                                mode="constant",
+                                constant_value=0,
+                                pad_width=(0, 0, 0, 0, pad_before, pad_after, 0, 0))
+            # (batch_size, num_filter, seq_len, num_scores=1)
+            conv = mx.sym.Convolution(data=padded,
+                                      #cudnn_tune="off",
+                                      kernel=(filter_width, self.num_embed),
+                                      num_filter=num_filter,
+                                      weight=self.conv_weight[filter_width],
+                                      bias=self.conv_bias[filter_width])
+            conv = mx.sym.Activation(data=conv, act_type="relu")
+            conv_outputs.append(conv)
+        # (batch_size, total_num_filters, seq_len, num_scores=1)
+        conv_concat = mx.sym.concat(*conv_outputs, dim=1)
+
+        # Max pooling with stride
+        uncovered = seq_len % self.pool_stride
+        if uncovered > 0:
+            pad_after = self.pool_stride - uncovered
+            # (batch_size, total_num_filters, seq_len + pad_to_final_stride, num_scores=1)
+            conv_concat = mx.sym.pad(data=conv_concat,
+                                     mode="constant",
+                                     constant_value=0,
+                                     pad_width=(0, 0, 0, 0, 0, pad_after, 0, 0))
+        # (batch_size, total_num_filters, seq_len/stride, num_scores=1)
+        pool = mx.sym.Pooling(data=conv_concat,
+                              pool_type="max",
+                              kernel=(self.pool_stride, 1),
+                              stride=(self.pool_stride, 1))
+        # (batch_size, total_num_filters, seq_len/stride)
+        pool = mx.sym.reshape(data=pool,
+                              shape=(-1, total_num_filters, encoded_seq_len))
+        # (batch_size, seq_len/stride, total_num_filters)
+        pool = mx.sym.swapaxes(data=pool, dim1=1, dim2=2)
+        if self.dropout > 0:
+            pool = mx.sym.Dropout(data=pool, p=self.dropout)
+
+        # Highway network
+        # (batch_size * seq_len/stride, total_num_filters)
+        pool_flat = mx.sym.Reshape(data=pool, shape=(-3, total_num_filters))
+        seg_embedding = pool_flat
```

Comment:
That's to be clear about each step.  The note that we're flattening the symbol could move to a comment.